manual/core/speculative_execution/README.md

```diff
@@ -0,0 +1,209 @@
+  |                |-------------------->|
+  |                |              |      |
+  |                |     server error    |
+  |                |<--------------------|
+  |                |              |
+  |                |   retry on node3
+  |                |-------------------->|
+  |                |              |      |
+  |                | node1 replies|      |
+  |                |<-------------|      |
+  |   complete     |                     |
+  |<---------------|                     |
+  |                | cancel              |
+  |                |-------------------->|
+```
+
+The only impact is that all executions of the same query always share the same query plan, so each 
+node will be used by at most one execution.
+
+### Tuning and practical details
+
+The goal of speculative executions is to improve overall latency (the time between `execute(query)`
+and `complete` in the diagrams above) at high percentiles. On the flip side, too many speculative
+executions increase the pressure on the cluster. 
+
+If you use speculative executions to avoid unhealthy nodes, a good-behaving node should rarely hit
+the threshold. We recommend running a benchmark on a healthy platform (all nodes up and healthy) and
+monitoring the request percentiles with the `cql-requests` [metric](../metrics/). Then use the
+latency at a high percentile (for example p99.9) as the threshold.
+
+Alternatively, maybe low latency is your absolute priority, and you are willing to take the
+increased throughput as a tradeoff. In that case, set the threshold to 0 and provision your cluster
+accordingly. 
+
+You can monitor the number of speculative executions triggered by each node with the
+`speculative-executions` [metric](../metrics/).
+
+#### Stream id exhaustion
+
+One side-effect of speculative executions is that many requests get cancelled, which can lead to a
+phenomenon called *stream id exhaustion*: each TCP connection can handle multiple simultaneous
+requests, identified by a unique number called *stream id* (see also the [pooling](../pooling/)
+section). When a request gets cancelled, we can't reuse its stream id immediately because we might
+still receive a response from the server later. If this happens often, the number of available
+stream ids diminishes over time, and when it goes below a given threshold we close the connection
+and create a new one. If requests are often cancelled, you will see connections being recycled at a
+high rate.
+
+The best way to monitor this is to compare the `pool.orphaned-streams` [metric](../metrics/) to the
+total number of available stream ids (which can be computed from the configuration:
+`pool.local.size * max-requests-per-connection`). The `pool.available-streams` and `pool.in-flight`
+metrics will also give you an idea of how many stream ids are left for active queries.
+
+#### Request ordering
+
+Note: ordering issues are only a problem with [server-side timestamps](../query_timestamps/), which
```

Comment:
Technically client timestamps have been the default since 2.1 (I think)

Reply:
Client timestamps are available in all the protocol versions supported by driver 4 (v3 and above).

Reply:
To clarify, my comment was more to the statement of `which are not the default anymore in driver 4+`.  This is technically true for driver 3+ right?